<launch>

    <!-- Args -->
    <arg name="publish_stereo_cloud" default="true"/>
    <arg name="enable_rviz" default="true"/>
    <arg name="from_bag" default="true"/>
    
    <!-- Do not change -->
    <param name="/use_sim_time" value="$(arg from_bag)" />
    <node pkg="lrm_localization_stack" type="imu_broadcaster_node" name="imu_broadcaster_node" output="screen"/>
    <node pkg="lrm_localization_stack" type="ins_to_utm_pose_node.py" name="ins_to_utm_pose_node" output="screen"/>
    <node pkg="lrm_localization_stack" type="ins_to_utm_pose_node.py" name="groundtruth_to_utm_pose_node" output="screen">
        <remap from="/carina/localization/nav_sat_fix" to="/carina/localization/nav_sat_fix_gt"/>
        <remap from="/carina/sensor/ins_utm_pose" to="/carina/sensor/ins_utm_pose_gt"/>
    </node>

    <!-- Visual odometry -->
    <arg name="visual_odometry_topic"   default="/localization/odometry/visual"/>
    <node pkg="rtabmap_ros" type="stereo_odometry" name="stereo_odometry" clear_params="true" args="" launch-prefix="">
        <remap from="left/image_rect"        to="/carina/sensor/camera/left/image_rect_color"/>
        <remap from="right/image_rect"       to="/carina/sensor/camera/right/image_rect_color"/>
        <remap from="left/camera_info"       to="/carina/sensor/camera/left/camera_info"/>
        <remap from="right/camera_info"      to="/carina/sensor/camera/right/camera_info"/>
        <remap from="odom"                   to="$(arg visual_odometry_topic)"/>
        <!-- <remap from="imu"                    to="$(arg imu_topic)"/> -->

        <param name="frame_id"                    type="string" value="base_link"/>
        <param name="odom_frame_id"               type="string" value="visual_odom"/>
        <param name="publish_tf"                  type="bool"   value="false"/>
        <param name="wait_for_transform_duration" type="double" value="0.2"/>
        <param name="wait_imu_to_init"            type="bool"   value="false"/>
        <param name="approx_sync"                 type="bool"   value="false"/>
        <param name="subscribe_rgbd"              type="bool"   value="false"/>
        <param name="guess_min_translation"       type="double" value="0"/>
        <param name="guess_min_rotation"          type="double" value="0"/>
        <param name="expected_update_rate"        type="double" value="0"/>
        <param name="max_update_rate"             type="double" value="0"/>
        <param name="keep_color"                  type="bool"   value="true"/>
    </node>
    <node if="$(arg publish_stereo_cloud)" pkg="nodelet" type="nodelet" name="stereo_cloud_publisher" args="standalone rtabmap_ros/point_cloud_xyzrgb" clear_params="true">
        <remap from="left/image"        to="/carina/sensor/camera/left/image_rect_color"/>
        <remap from="right/image"       to="/carina/sensor/camera/right/image_rect_color"/>
        <remap from="left/camera_info"  to="/carina/sensor/camera/left/camera_info"/>
        <remap from="right/camera_info" to="/carina/sensor/camera/right/camera_info"/>
        <remap from="cloud"             to="voxel_cloud" />

        <param name="decimation"  type="double" value="1"/>
        <param name="voxel_size"  type="double" value="0.1"/>
        <param name="approx_sync" type="bool"   value="false"/>
    </node>

    <!-- Fusion -->
    <node pkg="lrm_localization_stack" type="ekf_node" name="ekf_node" output="screen"/>

    <!-- Maintenance -->
    <node pkg="rosbag" type="play" name="player" args="--rate 0.3 --clock $(find lrm_localization_stack)/data/filtered_2023-01-17-12-28-28.bag --pause" if="$(arg from_bag)"/>
    <node type="rviz" name="rviz" pkg="rviz" args="-d $(find lrm_localization_stack)/config/localization.rviz" if="$(arg enable_rviz)" output="screen"/>

</launch>